breakend validation

yunfeiguo opened the pull request:
- [x] [RES-310](https://binatechnologies.atlassian.net/browse/RES-310)
- [x] Blocking PRs: ADD HERE IF ANY
* People
  - [x] Reviewers (>=1): @marghoob 
  - [x] Merger (>=1): @marghoob 
  - [x] FYI: @msahraeian, @ulitskyb
- [x] Release note (if needed)
- [x] No code copied from outside Bina 
- [x] Tests (Unit, Workflow, ITL & Gherkin) are passing
- [x] PR labels, milestones & assignees

# Summary

* Added an `Alt` class for storing alternative variants by not only sequence, but also symbolic allele and breakend. 
* Change `alts` field in `Variant` to `Alt` type. To maintain original behavior, `alts[i].getSeq()` is used to obtain the `FlexSeq` object. Later this should be improved.
* make breakend comparison aware of position, sidedness, orientation of mate.
* breakend is considered as interval of length 1 for looking for truth matches, mate is compared if two breakends overlap.
* 6 breakend/translocation tests added.

marghoob commented on src/main/java/com/bina/varsim/types/variant/alt/Alt.java:
```diff
@@ -0,0 +1,118 @@
+     * @param alt
+     */
+    public static SymbolicAllele SymbolicAlleleFactory(String alt) {
+      Pattern r = Pattern.compile("<([^<:>]+):?([^<:>]*)>");
```
This could be a final variable outside of the function so that this pattern does not need to be compiled every function call.

yunfeiguo replied:
Thanks! Fixed. Will continue working on other things.

marghoob commented on src/main/java/com/bina/varsim/util/VCFparser.java:
```diff
@@ -319,301 +319,286 @@ else if (index == 9) { // Output format
-                    }
-                }
-                //pos is incremented by 1, so it becomes 1-based start
+                      if (subtype[altAlleleIndex - 1].equals("ACCEPT")) {
```
Should make constant strings like these as constant variables defined in the class.

marghoob commented on src/main/java/com/bina/varsim/util/VCFparser.java:
```diff
@@ -319,301 +319,286 @@ else if (index == 9) { // Output format
-                //pos is incremented by 1, so it becomes 1-based start
+                      if (subtype[altAlleleIndex - 1].equals("ACCEPT")) {
+                          alts[altAlleleIndex - 1].setSeq(new FlexSeq(FlexSeq.Type.TRA, altAllelelength, currentCopyNumber));
+                      } else if (subtype[altAlleleIndex - 1].equals("REJECT")) {
```
Should make constant variables as earlier.

marghoob commented on src/main/java/com/bina/varsim/util/VCFparser.java:
```diff
@@ -319,301 +319,286 @@ else if (index == 9) { // Output format
+              if (svlens.length > 0) {
+                  for (int i = 0; i < svlens.length; i++) {
+                      int alternativeAlleleLength;
+                      if (svlens[i] == 0) {
```
Can probably just use a single conditional statement here using `? ... :` format.

marghoob commented on src/main/java/com/bina/varsim/util/StringUtilities.java:
```diff
@@ -51,10 +52,39 @@ public static String concatenateArray(final ChrString[] a, final CharSequence de
+    /**
+     * partial implementation of Java 8 StringJoiner
+     */
+    public static class StringJoiner {
```
Why do you need this? Isn't there already a StringJoiner in Java 8 and we are building using Java 8 anyway?

yunfeiguo replied:
Oh, I saw that on VarSim manual, java 7 was required. Shall we change it to Java 8?

marghoob replied:
I see. Yeah, we have moved to Java 8, primarily due to HTSJDK being Java 8 now. Please update the manual. Thanks.

yunfeiguo replied:
Sure. Thanks!

marghoob commented:
@yunfeiguo please do not forget to merge master to ensure no merge conflicts when the branch is finally ready for merging.

yunfeiguo commented:
@marghoob breakend validation is almost done, except that it needs more comprehensive testing.

marghoob commented on src/main/java/com/bina/varsim/tools/simulation/VCF2diploid.java:
```diff
@@ -739,7 +750,11 @@ private void writeVCF(final Sequence refSeq, final List<Variant> varList,
                         sbStr.append("TRASUBTYPE=");
+<<<<<<< HEAD
+                        sbStr.append(StringUtilities.concatenateArray(curr_var.getAllTranslocationSubtypeString(), ","));
+=======
```
Looks like there were merge conflicts. You will need to fix these manually.

yunfeiguo replied:
I just saw it. Thanks!

yunfeiguo commented:
@marghoob, could you please review and merge this PR? I have added a few more tests and fixed some issues in breakend validation. More tests are definitely needed however I want to change translocation format to the consensus in [RES-444](https://binatechnologies.atlassian.net/browse/RES-444) first. Thanks.

marghoob commented on src/main/java/com/bina/intervalTree/SimpleInterval1D.java:
```diff
@@ -186,6 +186,23 @@ public boolean intersects(Interval1D that, double reciprocalRatio) {
+//            */
+//            return this.getLeft() + wiggle >= that.getLeft() && this.getRight() - wiggle <= that.getRight();
         }
         long len_this = reciprocalRatio == 1.0 ? this.length() : (long) Math.max(Math.ceil(this.length() * reciprocalRatio), 1l);
```
The logic can be simplified here. Please have a look at https://github.com/bioinform/sandbox/blob/master/marghoob/circleseq/circleseq/utils.py#L59 . Of course, you will need to check that it also works for 0 length intervals.

yunfeiguo replied:
Are you suggesting something like

```
long len_this = (long) Math.max(Math.ceil(this.length() * reciprocalRatio), 1l);
```

? I will remove length 0 interval comparison since VarSim uses length as a way to keep track of comparison results of canonicalized variants (and it's weird that two length 0 intervals overlap).

yunfeiguo replied:
Refactored. Thanks!

marghoob commented on src/main/java/com/bina/varsim/tools/simulation/VCF2diploid.java:
```diff
@@ -441,11 +441,16 @@ private boolean addVariant(final byte[] maskedSequence, final Sequence reference
-                } else {
-                    insertions = allSequences.getSequence(variant.getChr2(allele)).revComp(variant.getEnd2(allele), variant.getPos2(allele) + 1);
-                }
+              if (Variant.TranslocationSubtype.ACCEPT.equals(variant.getTranslocationSubtype(allele))) {
```
Can't you use `Variant.TranslocationSubtype.ACCEPT == variant.getTranslocationSubtype(allele)` here?

marghoob commented on src/main/java/com/bina/varsim/tools/simulation/VCF2diploid.java:
```diff
@@ -478,17 +483,24 @@ private boolean addVariant(final byte[] maskedSequence, final Sequence reference
-                    s.setReferenceAlleleLength(variant.getReferenceAlleleLength());
+                FlexSeq s = null;
+
+                if (FlexSeq.Type.TRA.equals(variant.getAlt(allele).getSeqType())) {
```
Can't you use `FlexSeq.Type.TRA == variant.getAlt(allele).getSeqType()`?

marghoob commented on src/main/java/com/bina/varsim/tools/simulation/randVCFgenerator.java:
```diff
@@ -82,9 +83,9 @@ public byte sample_genotype(byte geno, Sample_params seen_added, int num_sample,
+        Alt alt = var.getAlt(geno);
         if (alt != null) {
-            if (alt.getType() == FlexSeq.Type.INS) {
+            if (FlexSeq.Type.INS.equals(alt.getSeqType())) {
```
Use `FlexSeq.Type.INS == alt.getSeqType()`?

yunfeiguo replied:
You are right. For `enum`, `==` and `equals()` are equivalent. I think this is more of my personal preference for comparing objects. I want to avoid `==` except for primitives.

marghoob commented:
Thanks. Is this good for merging now?

yunfeiguo commented:
Yes! Thanks!
